src/v/cluster/ephemeral_credential_frontend.cc

```diff
@@ -0,0 +1,179 @@
+
+#include "cluster/ephemeral_credential_frontend.h"
+
+#include "cluster/ephemeral_credential_serde.h"
+#include "cluster/ephemeral_credential_service_impl.h"
+#include "cluster/logger.h"
+#include "features/feature_table.h"
+#include "random/generators.h"
+#include "rpc/connection_cache.h"
+#include "security/acl.h"
+#include "security/credential_store.h"
+#include "security/ephemeral_credential.h"
+#include "security/ephemeral_credential_store.h"
+#include "security/exceptions.h"
+#include "security/scram_authenticator.h"
+#include "security/scram_credential.h"
+#include "security/types.h"
+#include "vformat.h"
+
+#include <seastar/coroutine/exception.hh>
+
+#include <fmt/ranges.h>
+
+// namespace
+
+namespace cluster {
+
+namespace {
+
+constexpr size_t credential_length{32};
+constexpr auto timeout = 5s;
+
+using scram = security::scram_sha512_authenticator::auth::scram;
+
+security::ephemeral_credential
+make_ephemeral_credential(security::acl_principal const& principal) {
+    constexpr auto gen = []() {
+        return random_generators::gen_alphanum_string(credential_length);
+    };
+    return {security::ephemeral_credential{
+      principal,
+      security::credential_user{gen()},
+      security::credential_password{gen()},
+      security::scram_sha512_authenticator::name}};
+}
+
+security::scram_credential
+make_scram_credential(security::ephemeral_credential const& cred) {
+    return scram::make_credentials(
+      cred.principal(), cred.password(), scram::min_iterations);
+}
+
+} // namespace
+
+using namespace std::chrono_literals;
+
+ephemeral_credential_frontend::ephemeral_credential_frontend(
+  model::node_id self,
+  ss::sharded<security::credential_store>& c_store,
+  ss::sharded<security::ephemeral_credential_store>& e_store,
+  ss::sharded<features::feature_table>& feature_table,
+  ss::sharded<rpc::connection_cache>& connections)
+  : _self(self)
+  , _c_store{c_store}
+  , _e_store{e_store}
+  , _feature_table{feature_table}
+  , _connections{connections}
+  , _gate{} {}
+
+ss::future<ephemeral_credential_frontend::get_return>
+ephemeral_credential_frontend::get(security::acl_principal const& principal) {
+    auto guard = _gate.hold();
+    get_return res;
+
+    if (!_feature_table.local().is_active(
+          features::feature::ephemeral_secrets)) {
+        vlog(
+          clusterlog.info,
+          "Ephemeral credentials feature is not active (upgrade in progress?)");
+        res.err = errc::invalid_node_operation;
+        co_return res;
+    }
+
+    if (auto it = _e_store.local().find(principal); !_e_store.local().has(it)) {
+        res.credential = make_ephemeral_credential(principal);
+        co_await put(res.credential);
+    } else {
+        res.credential = *it;
+    }
+
+    if (!_c_store.local().contains(res.credential.user())) {
+        co_await put(
+          res.credential.principal(),
+          res.credential.user(),
+          make_scram_credential(res.credential));
+    }
```

Comment:
it's not immediately obvious why the credential needs to go into the regular credential store here. presumably this code path is for the client, so making and storing an ephemeral credential makes sense. but then, adding to the normal store, would that imply that some broker is going to try to authenticate against _this_ broker that the client is on with the same credential?

Reply:
I've deleted and added this code back several times!

Currently It's so that it can be looked up for the inform call. But that's kinda nonsense, because it can be generated and there's no need to store it locally.

Reply:
OK, this breaks a test. Let me change the test to be what it should have tested in the first place; that the other broker can authenticate the ephemeral_credentials.

Reply:
cool. other than resolving this question this pr is pretty much ready to go

Reply:
Yay!

Reply:
It's a bit too complicated to test because I'd have to go through the whole scram dance due to not having the credentials that are sent, and the randomness of the salt.

Reply:
Actually, not that hard.